<a href="https://colab.research.google.com/github/VikiTarasova/DTA_2026/blob/main/project_store_sales_analysis/store_sales_analysis.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

 # Аналіз продажів магазину


# Вхідний датасет (дані)

In [34]:
# Запусти цей блок, щоб отримати датасет
import pandas as pd
import numpy as np

np.random.seed(42)
n = 300

categories = ['Молочне', 'Хліб та випічка', 'М\'ясо та риба', 'Овочі та фрукти', 'Напої']
products = {
    'Молочне':          ['Молоко 1л', 'Кефір 0.5л', 'Сир кисломолочний', 'Масло вершкове', 'Йогурт'],
    'Хліб та випічка':  ['Хліб білий', 'Батон', 'Булочка', 'Хліб житній', 'Круасан'],
    'М\'ясо та риба':   ['Куряче філе', 'Свинина', 'Ковбаса', 'Риба хек', 'Сосиски'],
    'Овочі та фрукти':  ['Картопля', 'Морква', 'Яблука', 'Банани', 'Помідори'],
    'Напої':            ['Вода 1.5л', 'Сік апельсиновий', 'Чай', 'Кава мелена', 'Лимонад'],
}
prices = {
    'Молоко 1л': 38, 'Кефір 0.5л': 22, 'Сир кисломолочний': 55, 'Масло вершкове': 95, 'Йогурт': 30,
    'Хліб білий': 28, 'Батон': 25, 'Булочка': 12, 'Хліб житній': 32, 'Круасан': 18,
    'Куряче філе': 165, 'Свинина': 210, 'Ковбаса': 145, 'Риба хек': 120, 'Сосиски': 98,
    'Картопля': 20, 'Морква': 15, 'Яблука': 45, 'Банани': 52, 'Помідори': 60,
    'Вода 1.5л': 18, 'Сік апельсиновий': 68, 'Чай': 85, 'Кава мелена': 155, 'Лимонад': 35,
}

chosen_cats = np.random.choice(categories, n, p=[0.25, 0.20, 0.20, 0.20, 0.15])
chosen_products = [np.random.choice(products[c]) for c in chosen_cats]

df_origin = pd.DataFrame({
    'transaction_id': range(1001, 1001 + n),
    'date':           pd.to_datetime(
                          np.random.choice(pd.date_range('2024-03-01', '2024-03-31'), n)
                      ),
    'hour':           np.random.choice(range(8, 21), n, p=[0.04,0.06,0.08,0.10,0.12,0.12,0.11,0.10,0.09,0.08,0.06,0.04,0.00]),
    'category':       chosen_cats,
    'product':        chosen_products,
    'quantity':       np.random.randint(1, 6, n),
    'price_uah':      [prices[p] for p in chosen_products],
    'customer_age':   np.random.randint(18, 70, n),
    'payment':        np.random.choice(['Готівка', 'Картка', 'Телефон'], n, p=[0.35, 0.50, 0.15]),
})


In [35]:
df = df_origin.copy()

In [36]:
df

,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment
0,1001,2024-03-17,17,Хліб та випічка,Круасан,4,18,28,Готівка
1,1002,2024-03-13,14,Напої,Кава мелена,3,155,40,Картка
2,1003,2024-03-01,18,Овочі та фрукти,Яблука,1,45,43,Картка
3,1004,2024-03-02,8,М'ясо та риба,Риба хек,3,120,51,Готівка
4,1005,2024-03-09,19,Молочне,Йогурт,5,30,63,Картка
...,...,...,...,...,...,...,...,...,...
295,1296,2024-03-08,10,М'ясо та риба,Сосиски,4,98,48,Готівка
296,1297,2024-03-05,13,Овочі та фрукти,Морква,5,15,28,Телефон
297,1298,2024-03-04,11,Молочне,Молоко 1л,3,38,27,Картка
298,1299,2024-03-24,14,М'ясо та риба,Ковбаса,1,145,20,Готівка


In [37]:
df["total_uah"] = df["price_uah"] * df["quantity"]

 ## Блок 1 - Перше знайомство з даними


1. Виведи перші 7 рядків датафрейму

In [38]:
# Виведи перші 7 рядків датафрейму
df.head(7)

,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
0,1001,2024-03-17,17,Хліб та випічка,Круасан,4,18,28,Готівка,72
1,1002,2024-03-13,14,Напої,Кава мелена,3,155,40,Картка,465
2,1003,2024-03-01,18,Овочі та фрукти,Яблука,1,45,43,Картка,45
3,1004,2024-03-02,8,М'ясо та риба,Риба хек,3,120,51,Готівка,360
4,1005,2024-03-09,19,Молочне,Йогурт,5,30,63,Картка,150
5,1006,2024-03-03,13,Молочне,Кефір 0.5л,3,22,51,Картка,66
6,1007,2024-03-31,15,Молочне,Масло вершкове,2,95,46,Телефон,190


 2. Дізнайся розмір таблиці (рядки × стовпці)


In [39]:
# Дізнайся розмір таблиці (рядки × стовпці).
print(f"Розмір: {df.shape[0]} рядків × {df.shape[1]} стовпців")

Розмір: 300 рядків × 10 стовпців


3. Виведи типи даних кожного стовпця


In [40]:
# Виведи типи даних кожного стовпця
print("Типи даних:")
df.dtypes

Типи даних:


,0
transaction_id,int64
date,datetime64[ns]
hour,int64
category,object
product,object
quantity,int64
price_uah,int64
customer_age,int64
payment,object
total_uah,int64


4. Перевір, чи є пропущені значення

In [41]:
# Перевір, чи є пропущені значення
df.isna()

,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
0,False,False,False,False,False,False,False,False,False,False
1,False,False,False,False,False,False,False,False,False,False
2,False,False,False,False,False,False,False,False,False,False
3,False,False,False,False,False,False,False,False,False,False
4,False,False,False,False,False,False,False,False,False,False
...,...,...,...,...,...,...,...,...,...,...
295,False,False,False,False,False,False,False,False,False,False
296,False,False,False,False,False,False,False,False,False,False
297,False,False,False,False,False,False,False,False,False,False
298,False,False,False,False,False,False,False,False,False,False


In [42]:
print(f"Попущені значення (по стовпцям):\n\n{df.isnull().sum()}")

Попущені значення (по стовпцям):

transaction_id    0
date              0
hour              0
category          0
product           0
quantity          0
price_uah         0
customer_age      0
payment           0
total_uah         0
dtype: int64


5. Виведи основну статистику (describe) для числових стовпців

In [43]:
print("Статистика:\n")
df.describe().round(2)

Статистика:



,transaction_id,date,hour,quantity,price_uah,customer_age,total_uah
count,300.00,300,300.00,300.00,300.00,300.0,300.00
mean,1150.50,2024-03-16 02:14:24,13.72,2.97,63.45,43.0,186.97
min,1001.00,2024-03-01 00:00:00,8.00,1.00,12.00,18.0,12.00
25%,1075.75,2024-03-08 00:00:00,12.00,2.00,22.00,31.0,60.00
50%,1150.50,2024-03-16 12:00:00,14.00,3.00,38.00,42.5,120.00
75%,1225.25,2024-03-24 00:00:00,16.00,4.00,95.75,55.0,256.25
max,1300.00,2024-03-31 00:00:00,19.00,5.00,210.00,69.0,840.00
std,86.75,NaN,2.74,1.44,51.68,14.6,180.27


**ВИСНОВОК:**

Датасет містить інформацію про 300 транзакцій, 9 стовпців, пропук значень немає.  
Середня сума чеку - 186.97грн.  
Мінімальний чек - 12.00грн ().  
Максимальний чек - 840.00грн ().  
Підозрілих даних не виявлено.

###### Розбір


```
[1, 7, 9, 10, 12, 15]
```
Медіана: (9 + 10) / 2 = 9,5
```
[1, 7, 9, 10, 12, 15, 20]
```
Медіана: 10

```
[1, 7, 9, 10, 12, 15, 20]
```
Медіана: 10 - `50%`  
`min` = 1  
`max` = 20  
`count` = 7  
`mean` = 10.57  
`25%` ->  
0% - min -> 25% -> 50% - median -> 75% -> 100% - max
std ->  
10.57 - 1 = 9.57
10.57 - 7 = 3.57
..
20 - 10.57 = 9.43

`[1, 1, 1, 1, 1, 6, 7, 8, 9, 10]`
1% - min -> 1  
25% -> 1  
50% - median -> 3.5  
75% -> 8  
100% - max -> 10

s = [1, 7, 9, 10, 12, 15, 20]
mean = sum(s)/len(s)
mean

s = [1, 1, 1, 1, 1, 1, 1000]  
mean = sum(s)/len(s)  
mean  
std -> 143.71 - 1 = 142.71  
mediana = 1


## Блок 2 - Фільтрація та відбір

1. Відбери всі покупки категорії «М'ясо та риба».

In [44]:
df[df['category'] == "М'ясо та риба"].head()

,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
3,1004,2024-03-02,8,М'ясо та риба,Риба хек,3,120,51,Готівка,360
8,1009,2024-03-21,15,М'ясо та риба,Ковбаса,2,145,34,Картка,290
17,1018,2024-03-06,14,М'ясо та риба,Куряче філе,3,165,69,Картка,495
20,1021,2024-03-23,9,М'ясо та риба,Куряче філе,3,165,64,Телефон,495
24,1025,2024-03-10,14,М'ясо та риба,Свинина,4,210,28,Картка,840


In [45]:
meat_fish = df[df['category'] == "М'ясо та риба"]
print(f'У категорії "М\'ясо та риба": {len(meat_fish)} транзакцій')

У категорії "М'ясо та риба": 61 транзакцій


2. Знайди транзакції, де сума (total_uah) перевищує 500 грн.

In [46]:
big = df[df['total_uah'] > 500]
print(f'Покупки > 500грн: {len(big)}')
print(big[['product', 'quantity', 'total_uah']])

Покупки > 500грн: 17
         product  quantity  total_uah
24       Свинина         4        840
28   Куряче філе         4        660
81       Ковбаса         4        580
89      Риба хек         5        600
93       Свинина         3        630
96       Свинина         3        630
101      Свинина         3        630
103  Куряче філе         5        825
146     Риба хек         5        600
147  Куряче філе         5        825
187      Свинина         4        840
193  Куряче філе         4        660
203  Кава мелена         5        775
221     Риба хек         5        600
241  Кава мелена         5        775
263     Риба хек         5        600
273  Кава мелена         4        620


3. Відбери покупки, зроблені карткою (payment == 'Картка'), кількістю більше 2 одиниць.

In [47]:
card_mult = df[(df['payment'] == 'Картка') & (df['quantity'] > 2)]

print("Картка + кількість > 2", len(card_mult),"\n")

card_mult.head()

Картка + кількість > 2 92 



,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
1,1002,2024-03-13,14,Напої,Кава мелена,3,155,40,Картка,465
4,1005,2024-03-09,19,Молочне,Йогурт,5,30,63,Картка,150
5,1006,2024-03-03,13,Молочне,Кефір 0.5л,3,22,51,Картка,66
10,1011,2024-03-23,13,Молочне,Сир кисломолочний,5,55,30,Картка,275
17,1018,2024-03-06,14,М'ясо та риба,Куряче філе,3,165,69,Картка,495


4. Знайди всі покупки покупців молодше 25 років.

In [48]:
young = df[df['customer_age'] < 25]
print("Покупців до 25: ", len(young))
young.head()

Покупців до 25:  36


,transaction_id,date,hour,category,product,quantity,price_uah,customer_age,payment,total_uah
7,1008,2024-03-01,17,Напої,Сік апельсиновий,1,68,21,Готівка,68
19,1020,2024-03-21,16,Хліб та випічка,Круасан,1,18,19,Готівка,18
26,1027,2024-03-19,13,Молочне,Йогурт,3,30,20,Готівка,90
36,1037,2024-03-04,14,Хліб та випічка,Круасан,1,18,23,Телефон,18
38,1039,2024-03-01,10,Овочі та фрукти,Помідори,1,60,22,Готівка,60


5. Скільки унікальних товарів є в датасеті?

In [49]:
n_unique = df['product'].nunique() # кількість унікальних товарів
n_unique

pr_unique = df['product'].unique() # перелік унікальних товарів
pr_unique

print("кількість унікальних товарів", n_unique, "\n")
print("перелік унікальних товарів:\n", pr_unique)

кількість унікальних товарів 25 

перелік унікальних товарів:
 [np.str_('Круасан') np.str_('Кава мелена') np.str_('Яблука')
 np.str_('Риба хек') np.str_('Йогурт') np.str_('Кефір 0.5л')
 np.str_('Масло вершкове') np.str_('Сік апельсиновий') np.str_('Ковбаса')
 np.str_('Картопля') np.str_('Сир кисломолочний') np.str_('Морква')
 np.str_('Куряче філе') np.str_('Хліб житній') np.str_('Батон')
 np.str_('Хліб білий') np.str_('Свинина') np.str_('Сосиски')
 np.str_('Молоко 1л') np.str_('Помідори') np.str_('Лимонад')
 np.str_('Булочка') np.str_('Банани') np.str_('Вода 1.5л') np.str_('Чай')]


відсоток великих покупок

In [50]:
pct_big = len(big) / len(df) * 100
# round(pct_big, 1)
print(f"Частка покупок > 500: {pct_big:.1f}")

Частка покупок > 500: 5.7


In [51]:
for cat in big.category.unique():
  # print(f"{cat} - {big['category'].count()}")
  print(f"{cat} - {big['category'].count()}")
  # print(f"{cat}")


М'ясо та риба - 17
Напої - 17


**ВИСНОВОК:**
Датасет містить інформацію про 300 транзакцій, 10 стовпців, пропущених значень немає.
Середня сума чеку - 186.97 грн.
Мінімальний чек - 12.00 грн ().
Максимальний чек - 840.00 грн ().
Підозрілих даних не виявлено.


## Блок 3 - Групування та агрегація

1. Підрахуй загальний виторг по кожній категорії. Відсортуй за спаданням.



In [52]:
# revenue_by_cat =
df.groupby('category')['total_uah'].sum().sort_values(ascending=False)

,total_uah
category,
М'ясо та риба,25223
Напої,10389
Молочне,10075
Овочі та фрукти,6270
Хліб та випічка,4134


In [53]:
df.groupby('category').agg({
   'total_uah': 'sum',
   'quantity': 'sum',
   'transaction_id': 'count'

}).sort_values(by="total_uah", ascending=False)

,total_uah,quantity,transaction_id
category,,,
М'ясо та риба,25223,186,61
Напої,10389,129,48
Молочне,10075,242,80
Овочі та фрукти,6270,158,55
Хліб та випічка,4134,177,56



2. Знайди топ-5 товарів за кількістю проданих одиниць (сума quantity).


In [57]:
print("Тор-5 продуктів за кількістю проданих одиниць:\n")
  #  top_products =
df.groupby('product')['quantity'].sum().sort_values(ascending=False).head(5)

Тор-5 продуктів за кількістю проданих одиниць:



,quantity
product,
Кефір 0.5л,63
Сосиски,61
Йогурт,57
Сир кисломолочний,54
Риба хек,48



3. Порівняй середню суму покупки для різних способів оплати.


In [59]:
print("Середній чек по оплаті:\n")
df.groupby("payment")['total_uah'].mean().round(2)

Середній чек по оплаті:



,total_uah
payment,
Готівка,175.16
Картка,205.26
Телефон,144.82



4. Згрупуй по hour - знайди, в які години відбувається найбільше транзакцій.


In [64]:
df.groupby('hour').size()

,0
hour,
8,7
9,11
10,23
11,30
12,30
13,41
14,44
15,27
16,32


In [63]:
df.groupby('hour')['hour'].count()

,hour
hour,
8,7
9,11
10,23
11,30
12,30
13,41
14,44
15,27
16,32


In [66]:
print("Транзакції по годинах:\n")
df['hour'].value_counts().sort_index()

Транзакції по годинах:



,count
hour,
8,7
9,11
10,23
11,30
12,30
13,41
14,44
15,27
16,32


In [67]:
print("Транзакції по годинах (за кількістю транзакцій):\n")
df['hour'].value_counts()

Транзакції по годинах (за кількістю транзакцій):



,count
hour,
14,44
13,41
16,32
12,30
11,30
15,27
17,26
10,23
18,19



5. Знайди день місяця з найвищим виторгом.

In [70]:
df.groupby('date')['total_uah'].sum()

,total_uah
date,
2024-03-01,1355
2024-03-02,3371
2024-03-03,2345
2024-03-04,1651
2024-03-05,2311
2024-03-06,2485
2024-03-07,845
2024-03-08,1593
2024-03-09,1628


In [75]:
df['date'].dt.day

,date
0,17
1,13
2,1
3,2
4,9
...,...
295,8
296,5
297,4
298,24


In [77]:
df['day'] = df['date'].dt.day
best_day = df.groupby('day')['total_uah'].sum().idxmax()
best_revenue = df.groupby('day')['total_uah'].sum().max()

print(f"Найкращий день: {best_day} березня - {best_revenue:.2f} грн")

Найкращий день: 2 березня - 3371.00 грн


**ВИСНОВОК**  
Найприбутковіша категорія - "М'ясо та риба" (висока ціна)  
Найчастіше купують - молочну продукцію, особливо кефір.  
Пікові години торгівлі - з 13:00 до 14:00, проте прирост клієнтів починається вже з 11:00.  
Покупці, що здійснюють транзакції за допомогою картки, мають в середньому більший чек ніж ті, хто оплачує готівкою.


## Блок 4 - Нові стовпці та трансформація

1. Додай стовпець revenue_share - частка виторгу кожної транзакції від загального виторгу (у відсотках).


2. Додай стовпець age_group:
'Молодь' - до 30 років
'Середній вік' - 30–50 років
'Старші' - понад 50 років

3. Додай стовпець weekday - день тижня з колонки date.

4. одай булевий стовпець is_big_purchase - True, якщо total_uah > 300.